### PR TITLE
Added mlang tags support to collections and items

### PR DESCRIPTION
**Pull request by Lenetis**
Item and collection names, field titles, field contents, sorting columns, and quite a few other things now support the {mlang} tags.

Example usage:
`Text that will always be displayed{mlang en} text in english{mlang}{mlang pl} tekst po polsku{mlang} always displayed{mlang en} another text in english{mlang}.`

### CODE REVIEW

**Review by punfil — COMMENTED** [review]
Looks pretty good, that's a cool feature.

Detected bug:
http://192.168.22.20:81/collections/9155d0a3-1833-47f6-b8f5-7de3a1b33216/items
obraz

bug2 (locale is not properly checked for typos/wrong input):
obraz
obraz

bug3:
obraz


Let's suppose I have sth like this (scenario where sth is not translated into all languages) :
{mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang}
How to display sth in all languages besides en and pl ^?
I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.

**Comment by Lenetis** [comment]
Bug 1: fixed
Bug 2: I don't really see a problem with that. Besides, what could be the solution?
Bug 3: fixed

> Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.

That's a good point. I have two ideas:
1. If not a single tag was matched, display english text by default
2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched

The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?

**Review by punfil — COMMENTED** [review]
Mostly LGTM, but need to retest once this branch is rebased to dev

**Comment by punfil** [comment]
> Bug 1: fixed Bug 2: I don't really see a problem with that. Besides, what could be the solution? Bug 3: fixed
> 
> > Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.
> 
> That's a good point. I have two ideas:
> 
>     1. If not a single tag was matched, display english text by default
> 
>     2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched
> 
> 
> The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?



> Bug 1: fixed Bug 2: I don't really see a problem with that. Besides, what could be the solution? Bug 3: fixed
> 
> > Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.
> 
> That's a good point. I have two ideas:
> 
>     1. If not a single tag was matched, display english text by default
> 
>     2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched
> 
> 
> The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?

The first one makes more sense to me.
I mean, with that we force user to translate item * to english, BUT probably if we declared mlang other it would contain english anyways.


Second thoughts, what about a setting which tag is the default language for not translated? That might be difficult though.

**Review by bartoszLesniewski — CHANGES_REQUESTED** [review]
Generally, mlang tags works correctly, but we MUST define some default behaviour when translation for user language doesn't exist. What about using both options that you mentioned?
I mean:
1. User can add `default` parameter in mlang tag, e.g. `{mlang en default}Some text{mlang}`, or even use separate mlang tag (`{mlang default}Some text{mlang}`) if handling two parameters in one tag is too difficult.
2. If default is not specified, english should be used by default

One more suggestion (if it would take a lot of work, you can skip it):
It is possible that user types more spaces between `mlang` and language shortcut or even between `mlang` and `{` by mistake, e.g. `{mlang  pl}Tekst{mlang}` or `{ mlang pl }Tekst{ mlang }`. For now, such cases aren't handled correctly. Is it possible to do it? 

Also, please update other places in source code where mlangs should be used after rebase to dev

**Comment by Lenetis** [comment]
> 1. User can add `default` parameter in mlang tag, e.g. `{mlang en default}Some text{mlang}`, or even use separate mlang tag (`{mlang default}Some text{mlang}`) if handling two parameters in one tag is too difficult.

The default attribute is a good idea. This is what I ended up with:

- `{mlang en}` - will be displayed only if the locale is set to english
- `{mlang default}` - will be displayed only if no other non-default tag was matched
- `{mlang en default}` - will be displayed if the locale is set to english OR if no other non-default tag was matched

If there are multiple `default` tags, all of them will be displayed if not a single non-default tag was matched.

-----

> 2. If default is not specified, english should be used by default

I don't think displaying english is necessary if we have the default attribute. There are valid situations where some text is outside the mlang tags, and they are used e.g. only as a clarification visible in one language

-----

> One more suggestion (if it would take a lot of work, you can skip it): It is possible that user types more spaces between `mlang` and language shortcut or even between `mlang` and `{` by mistake, e.g. `{mlang pl}Tekst{mlang}` or `{ mlang pl }Tekst{ mlang }`. For now, such cases aren't handled correctly. Is it possible to do it?

Done

-----
 
> Also, please update other places in source code where mlangs should be used after rebase to dev

So labels and Dublin Core, or is there something else besides these two?

File: assets/controllers/select/default_controller.js [hunk 1]
```diff
@@ -1,17 +1,29 @@
 import { Controller } from '@hotwired/stimulus';
 import Translator from "bazinga-translator";
 import { TsSelect2 } from "../../node_modules/ts-select2/dist/core";
-import { htmlStringToDomElement } from "../../js/utils";
```
Comment (punfil) [inline assets/controllers/select/default_controller.js]: rebase bug?

File: assets/controllers/select/default_controller.js [hunk 2]
```diff
@@ -1,17 +1,29 @@
...
 
+/* stimulusFetch: 'lazy' */
 export default class extends Controller {
     select2;
 
     connect() {
-        this.loadSelect()
-    }
```
Comment (punfil) [inline assets/controllers/select/default_controller.js]: Why change the scheme? The less we change the less conflicts
Please consider

File: templates/App/Collection/show.html.twig [hunk 3]
```diff
@@ -105,16 +105,16 @@
                         {% if collection.data is not empty %}
                             {% set underlinedTags = getUnderlinedTags(collection.data) %}
                             {% for datum in collection.data %}
-                                <div class="datum-row s12 l5 xl4">
-                                    <span class="label">{{ datum.label }} :</span>
+                                <div class="datum-row col s12 l5 xl4">
```
Comment (punfil) [inline templates/App/Collection/show.html.twig]: we need div class `col` here or rebase issue?
Same below

Please consider

**Review by punfil — CHANGES_REQUESTED** [review]
Found bug:
printable-list
obraz

and possible bug:
Export as CSV in Tools:
obraz

Just want to bring attention here, it exports {mlang}... but in other exports like DublinCore we use transMlang filter.

Did you check what happens when scrapers are used?

**Review by punfil — COMMENTED** [review]
Bug 3:
obraz

"Ekwipunek"

**Review by punfil — COMMENTED** [review]
Possible bug (Please consider)
Events history:
obraz

**Review by punfil — COMMENTED** [review]
Bug 4:
obraz

**Review by punfil — COMMENTED** [review]
I don't know if we want to support it here though:
obraz

**Review by punfil — COMMENTED** [review]
Bug 6:
obraz

**Review by punfil — COMMENTED** [review]
I don't know if we want to support it here though (album):
obraz

**Review by punfil — COMMENTED** [review]
I don't know if we want to support it here though:

obraz

**Review by punfil — COMMENTED** [review]
obraz
Adding related items tip shows:
{mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang}{mlang default}DEFAULTEKST{mlang}
instead of PLTEKST

**Review by punfil — APPROVED** [review]
LGTM, thanks for implementing the feedback!

**Review by bartoszLesniewski — APPROVED** [review]

**Pull request merged**